dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java

```diff
@@ -854,6 +890,83 @@ public Knowledge visitSameAddress(Relation rel) {
+            for (Event e1 : events) {
+                for (Event e2 : events) {
+                    ScopedThread thread1 = (ScopedThread) e1.getThread();
+                    ScopedThread thread2 = (ScopedThread) e2.getThread();
+                    if (specificScope != null) { // scope specified
+                        if (thread1.sameAtHigherScope(thread2, specificScope) && !exec.areMutuallyExclusive(e1, e2)) {
+                            must.add(new Tuple(e1, e2));
+                        }
+                    } else {
+                        String scope1 = Tag.PTX.getScopeTag(e1);
+                        String scope2 = Tag.PTX.getScopeTag(e2);
+                        if (scope1.equals(scope2) && !scope1.isEmpty() && thread1.sameAtHigherScope(thread2, scope1)
+                                && !exec.areMutuallyExclusive(e1, e2)) {
+                            must.add(new Tuple(e1, e2));
+                        }
+                    }
+                }
+            }
+            return new Knowledge(must, new HashSet<>(must));
+        }
+
+        @Override
+        public Knowledge visitSyncBarrier(Relation sync_bar) {
+            Set<Tuple> may = new HashSet<>();
+            Set<Tuple> must = new HashSet<>();
+            List<FenceWithId> fenceEvents = program.getEvents(FenceWithId.class);
+            for (FenceWithId e1 : fenceEvents) {
+                for (FenceWithId e2 : fenceEvents) {
+                    if(exec.areMutuallyExclusive(e1, e2)) {
+                        continue;
+                    }
+                    may.add(new Tuple(e1, e2));
+                    if (e1.getFenceID().equals(e2.getFenceID())) {
+                        must.add(new Tuple(e1, e2));
+                    }
+                }
+            }
+            return new Knowledge(may, must);
+        }
+
+        @Override
+        public Knowledge visitSyncFence(Relation sync_fen) {
+            Set<Tuple> may = new HashSet<>();
+            List<Fence> fenceEvents = program.getEvents(Fence.class);
+            for (Fence e1 : fenceEvents) {
+                for (Fence e2 : fenceEvents) {
+                    if (e1.is(Tag.PTX.SC) && e2.is(Tag.PTX.SC) && !exec.areMutuallyExclusive(e1, e2)) {
+                        may.add(new Tuple(e1, e2));
+                    }
+                }
+            }
+            return new Knowledge(may, EMPTY_SET);
+        }
+
+        @Override
+        public Knowledge visitAlias(Relation rel) {
+            Set<Tuple> must = new HashSet<>();
+            List<MemEvent> events = program.getEvents(MemEvent.class);
+            for (MemEvent e1 : events) {
+                for (MemEvent e2 : events) {
+                    if (samePhysicalAddress(e1, e2) && !exec.areMutuallyExclusive(e1, e2)) {
+                        must.add(new Tuple(e1, e2));
+                    }
+                }
+            }
+            return new Knowledge(must, new HashSet<>(must));
```

Comment:
I would change this definition with the definition of `loc`. I think `loc` should be based on the physical address while `alias` is based on the virtual one.

Reply:
Updated to `visitLoc`, but I'm afraid is it a bit confusing for Alias.java class to visitLoc?

Reply:
I'm not saying that you should just rename the function (that is indeed confusing).
I'm talking more generally that the semantics of `loc` and `alias` should be swapped around.
This also implies that you need to change the cat file and swap all occurences of `loc` and `alias`.
I've written a post about why this is better (just scroll up). But in short: this aligns better with the typical notion of `loc` in other models, it allows you to preserve the invariants `rf \subseteq loc` and `co \subseteq loc`, and it better aligns with the Alloy model!

Reply:
The `same_physical_address` has been moved into the `VirtualAliasAnalysis` (meanwhile removed from `co`, `rf` relations). The usage of `loc` and `alias`(now named `vloc`) in cat are swapped.  The `loc` relation now signifies the same physical address, whereas the `vloc` relation represents the same virtual address, which is stricter than `loc`.

Reply:
Perfect, that's what I had in mind.